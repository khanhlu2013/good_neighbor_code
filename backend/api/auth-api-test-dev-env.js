const route = require("express").Router();

const keys = require("../configs/keys");
const { FRONTEND_URL } = keys;

if (keys.NODE_ENV === "production") {
  throw Error("Unexpect production env");
}

route.get("/dummy_login_for_test_purpose", (req, res, next) => {
  const User = require("../models/user");
  const { email, name } = req.query;

  (async () => {
    const user = await User.findOneOrCreate(email, name);
    req.login(user, err => {
      if (err) {
        console.log(err);
        res.status(500).send();
      } else {
        return res.redirect(FRONTEND_URL);
      }
    });
  })().catch(next);
});

module.exports = route;
