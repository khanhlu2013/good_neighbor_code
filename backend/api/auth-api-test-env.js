const route = require('express').Router();

const keys = require('../configs/keys');
const {FRONTEND_URL} = keys;

if(keys.NODE_ENV !== 'test'){
  throw Error(`Unexpected ${keys.NODE_ENV} env`);
}

route.get('/dummy_login_for_test_purpose',async (req,res)=>{
  const User =  require('../models/user');
  const {email,name} = req.query;
  try{
    const user = await User.findOneOrCreate(email,name);
    req.login(user,(err)=>{
      if(err){
        console.log(err);
        res.status(500).send();
      }else{
        return res.redirect(FRONTEND_URL);
      }
    });
  }catch(e){
    console.log(e);
    res.status(500).send();
  }
});

module.exports = route;