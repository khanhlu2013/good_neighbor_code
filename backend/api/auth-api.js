const route = require("express").Router();
const passport = require("passport");

const keys = require("../configs/keys");

route.get(
  "/google",
  passport.authenticate("google", { scope: ["profile", "email"] })
);
route.get("/google/redirect", passport.authenticate("google"), (req, res) => {
  res.redirect(keys.FRONTEND_URL);
});

if (keys.NODE_ENV !== "production") {
  route.get("/login_for_test_dev", (req, res, next) => {
    const User = require("../models/user");
    const { email, name } = req.query;
    (async () => {
      const user = await User.findOneOrCreate(email, name);
      req.login(user, err => {
        if (err) {
          console.log(err);
          res.status(500).send();
        } else {
          return res.redirect(keys.FRONTEND_URL);
        }
      });
    })().catch(next);
  });
}

module.exports = route;
