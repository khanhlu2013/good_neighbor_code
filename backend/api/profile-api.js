const route = require("express").Router();
const mongoose = require("mongoose");

const Connection = require("../models/connection");
const Post = require("../models/post");
const User = require("../models/user");
const ShareLog = require("../models/shareLog");
const keys = require("../configs/keys");

// - auth ----
const authCheck = (req, res, next) => {
  if (!req.user) {
    return res.status(401).send();
  }
  next();
};

route.get("/logout", authCheck, (req, res) => {
  req.logout();
  res.redirect(keys.FRONTEND_URL);
});

route.get("/", authCheck, (req, res) => {
  res.send(req.user);
});

// - user ----
route.get("/searchEmail", authCheck, (req, res, next) => {
  const { email } = req.query;
  const { user } = req;

  if (!email) {
    return res.status(400).send();
  }

  if (email === user.email) {
    return res.status(400).send();
  }

  (async () => {
    res.send(await User.findOne({ email }));
  })().catch(next);
});

// - connection ----
route.get("/connections", authCheck, (req, res, next) => {
  const { user } = req;
  (async () => {
    const connections = await Connection.find({
      $or: [{ from: user }, { to: user }]
    })
      .populate("from")
      .populate("to");
    res.send(connections);
  })().catch(next);
});

route.post("/createConnection", authCheck, (req, res, next) => {
  const { user } = req;
  const { userIdToAdd } = req.body;

  if (!userIdToAdd) {
    return res.status(400).send();
  }

  if (!mongoose.Types.ObjectId.isValid(userIdToAdd)) {
    return res.status(400).send();
  }

  (async () => {
    const connection = await new Connection({
      from: user.id,
      to: userIdToAdd
    }).save();
    return res.send(connection);
  })().catch(next);
});

route.post("/updateConnection", authCheck, (req, res, next) => {
  const { user } = req;
  const { connectionId, isApproved } = req.body;

  if (connectionId === undefined || isApproved === undefined) {
    return res.status(400).send();
  }

  if (!mongoose.Types.ObjectId.isValid(connectionId)) {
    return res.status(400).send();
  }
  (async () => {
    const connection = await Connection.findById(connectionId);
    if (connection.from.equals(user._id)) {
      connection.approvedByFrom = isApproved;
    } else {
      connection.approvedByTo = isApproved;
    }
    return res.send(await connection.save());
  })().catch(next);
});

// - post ----
route.get("/posts", authCheck, (req, res, next) => {
  const { user } = req;
  const { isActive } = req.query;
  (async () => {
    const query = {
      by: user
    };
    if (isActive) {
      query.isActive = isActive;
    }
    const posts = await Post.find(query);

    res.send(posts);
  })().catch(next);
});

route.post("/createPost", authCheck, (req, res, next) => {
  const { user } = req;
  const { title, description } = req.body;
  (async () => {
    const post = new Post({
      by: user,
      title: title,
      description: description
    });
    await post.save();
    res.send(post);
  })().catch(next);
});

route.post("/updatePost", authCheck, (req, res, next) => {
  const { user } = req;
  const { postID, title, description } = req.body;

  (async () => {
    if (!postID) {
      return res.status(400).send();
    }

    if (!mongoose.Types.ObjectId.isValid(postID)) {
      return res.status(400).send();
    }

    const post = await Post.findById(postID);
    if (!post) {
      return res.status(400).send();
    }

    if (post.by.equal(user._id)) {
      return res.status(401).send();
    }

    Object.assign(post, { title, description });
    await post.save();
    res.send(post);
  })().catch(next);
});

// - shareLog
route.get("/shareLogs", authCheck, (req, res, next) => {
  const { isIn, isOut } = req.query;
  const { user } = req;

  if (!isIn && !isOut) {
    return res.status(400).send();
  }
  const query = {};
  if (isIn) {
    query.isIn = isIn;
  }
  if (isOut) {
    query.isOut = isOut;
  }

  (async () => {
    const shareLogs = await ShareLog.find(query);
    res.send(shareLogs);
  })().catch(next);
});

route.post("/createShareLog", authCheck, (req, res, next) => {
  const { user } = req;
  const { postID } = req.body;

  if (!mongoose.Types.ObjectId.isValid(postID)) {
    return res.status(400).send();
  }

  (async () => {
    const shareLog = new ShareLog({
      post: postID,
      borrower: user.id
    });
    await post.shareLog().save();
    res.send(shareLog);
  })().catch(next);
});

route.post("/updateInShareLog", authCheck, (req, res, next) => {
  const { user } = req;
  const { shareLogID, isReturning } = req.body;

  if (!mongoose.Types.ObjectId.isValid(shareLogID)) {
    return res.status(400).send();
  }

  (async () => {
    const shareLog = await ShareLog.findById(shareLogID);
    if (!shareLog) {
      return res.status(400).send();
    }

    if (!shareLog.borrower.equal(user._id)) {
      return res.status(401).send();
    }

    shareLog.isReturning = isReturning;
    await shareLog.save();
    res.send(shareLog);
  })().catch(next);
});

route.post("/updateOutShareLog", authCheck, (req, res, next) => {
  const { user } = req;
  const { shareLogID, isReturned } = req.body;

  if (!mongoose.Types.ObjectId.isValid(shareLogID)) {
    return res.status(400).send();
  }

  (async () => {
    const shareLog = await ShareLog.findById(shareLogID).populate("post");
    if (!shareLog) {
      return res.status(400).send();
    }

    if (!shareLog.post.by.equal(user._id)) {
      return res.status(401).send();
    }

    shareLog.isReturned = isReturned;
    await shareLog.save();
    res.send(shareLog);
  })().catch(next);
});

module.exports = route;
