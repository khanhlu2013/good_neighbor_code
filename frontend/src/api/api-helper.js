import { Connection } from "../model/connection";
import { User } from "../model/user";
import { Post } from "../model/post";
import { Share } from "../model/share";

const rawToConnection = raw => {
  const {
    _id: id,
    from: { _id: fromID, email: fromEmail, name: fromName },
    to: { _id: toID, email: toEmail, name: toName },
    approvedByTo,
    approvedByFrom
  } = raw;

  const from = new User(fromID, fromEmail, fromName);
  const to = new User(toID, toEmail, toName);
  return new Connection(id, from, to, approvedByTo, approvedByFrom);
};

const rawsToConnections = raws => {
  return raws.map(raw => rawToConnection(raw));
};

function rawsToPosts(raws) {
  return raws.map(raw => {
    const { post: postRaw, user: userRaw } = raw;
    const sharesRaw = raw.shares.filter(share => share._id !== undefined);

    const shares = sharesRaw.map(shareRaw => {
      const { _id: userId, email, name } = shareRaw.borrower;
      const borrower = new User(userId, email, name);
      return new Share(
        shareRaw._id,
        borrower,
        shareRaw.dateCreated,
        shareRaw.isApprovedByFrom,
        shareRaw.isReturnedByTo,
        null //post to be set later
      );
    });
    const { _id: userId, email: userEmail, name: userName } = userRaw;
    const post = new Post(
      postRaw._id,
      new User(userId, userEmail, userName),
      postRaw.isActive,
      postRaw.title,
      postRaw.description,
      postRaw.dateCreated,
      shares
    );
    for (const share of post.shares) {
      share.setPost(post);
    }
    return post;
  });
}

export { rawsToConnections, rawToConnection, rawsToPosts };
