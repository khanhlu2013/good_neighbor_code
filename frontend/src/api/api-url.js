import RouteNode from "route-node";

const { NODE_ENV } = process.env;
let BACKEND_BASE_URL;
let BACKEND_PORT;

if (NODE_ENV === "production") {
  BACKEND_BASE_URL = "https://goodneighbor-backend-test.herokuapp.com";
  BACKEND_PORT = "";
} else {
  BACKEND_BASE_URL = "http://localhost";
  if (process.env.REACT_APP_NODE_ENV === "test") {
    BACKEND_PORT = ":3001";
  } else {
    BACKEND_PORT = ":3000";
  }
}

const BACKEND_URL = `${BACKEND_BASE_URL}${BACKEND_PORT}`;

const API_NODE = new RouteNode("", "", [
  new RouteNode("profile", "/profile", [
    { name: "authCheck", path: "/authCheck" },
    { name: "logout", path: "/logout" },
    { name: "searchEmail", path: "/searchEmail?:email" },
    // - connection
    { name: "connections", path: "/connections" },
    { name: "createConnection", path: "/createConnection" },
    { name: "updateConnection", path: "/updateConnection" },
    // - post
    { name: "createPost", path: "/createPost" },
    { name: "updatePost", path: "/updatePost" },
    { name: "awareReturnPost", path: "/awareReturnPost" },
    // - outPosts
    { name: "outPosts", path: "/outPosts" },
    // - inPosts
    { name: "inPosts", path: "/inPosts" },
    // - share
    { name: "share", path: "/share" },
    { name: "createShare", path: "/createShare" },
    { name: "deleteShare", path: "/deleteShare" },
    { name: "approveShare", path: "/approveShare" },
    { name: "returnShare", path: "/returnShare" },
    { name: "awareApproveShare", path: "/awareApproveShare" }
  ]),
  new RouteNode("auth", "/auth", [
    { name: "google", path: "/google" },
    {
      name: "backdoorLogin",
      path: "/backdoorLogin?:email?:name"
    }
  ])
]);

const API_URL = (dottedPath, params) => {
  const path = API_NODE.buildPath(dottedPath, params);
  if (!path) {
    throw Error(`${dottedPath} is not found!`);
  }

  return BACKEND_URL + path;
};

export { API_URL };
