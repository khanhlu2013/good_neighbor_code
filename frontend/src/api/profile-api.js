import keys from "../configs/keys.js";

const profile = async () => {
  const request = await fetch(keys.API_URL("profile"), {
    credentials: "include"
  });
  if (request.status === 401) {
    return null;
  }
  return await request.json();
};

const searchEmail = email => {
  return get("profile.searchEmail", { email });
};

const connections = () => {
  return get("profile.connections", {});
};

const modifyConnection = (connectionId, isApproved) => {
  return post("profile.modifyConnection", {
    connectionId,
    isApproved
  });
};

const createConnection = userIdToAdd => {
  return post("profile.createConnection", {
    userIdToAdd
  });
};

const posts = async () => {
  return get("profile.posts", {});
};

const crudPost = async (postID, title, description) => {
  const crudedPost = await post("profile.crudPost", {
    postID,
    title,
    description
  });
  return crudedPost;
};

//- helper ----

async function getJSON(response) {
  const text = await response.text();
  if (!text) {
    return null;
  }
  return JSON.parse(text);
}

async function post(dottedPath, params) {
  const response = await fetch(keys.API_URL(dottedPath), {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(params),
    credentials: "include"
  });

  return await getJSON(response);
}

async function get(dottedPath, params) {
  const url = keys.API_URL(dottedPath, params);
  const response = await fetch(url, { credentials: "include" });
  return await getJSON(response);
}

//-------------

const API = {
  profile,
  searchEmail,
  connections,
  modifyConnection,
  createConnection,
  posts,
  crudPost
};
export { API };
