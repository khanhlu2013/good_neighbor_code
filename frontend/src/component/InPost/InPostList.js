import React from "react";
import PropTypes from "prop-types";
import { InPostItem } from "./inPostItem";

function InPostList(props) {
  const {
    posts,
    requestingPostIds,
    deletingShareIds,
    awaringShareIds,
    returningShareIds,
    onCreateShare,
    onDeleteShare,
    onAwareShare,
    onReturnShare,
    loginUser
  } = props;

  let content;
  if (posts.length === 0) {
    content = <p className="text-muted text-center h4">there are no data</p>;
  } else {
    content = posts
      .sort((p1, p2) => p2.dateCreated - p1.dateCreated)
      .map(post => (
        <InPostItem
          key={post.id}
          loginUser={loginUser}
          post={post}
          onCreateShare={onCreateShare}
          onDeleteShare={onDeleteShare}
          onAwareShare={onAwareShare}
          onReturnShare={onReturnShare}
          isRequestingPost={requestingPostIds.includes(post.id)}
          isDeleteingShare={post.shares.some(share =>
            deletingShareIds.includes(share.id)
          )}
          isAwaringShare={post.shares.some(share =>
            awaringShareIds.includes(share.id)
          )}
          isReturningShare={post.shares.some(share =>
            returningShareIds.includes(share.id)
          )}
        />
      ));
  }

  return <div id="InPostList-react">{content}</div>;
}
InPostList.propTypes = {
  posts: PropTypes.array.isRequired,
  requestingPostIds: PropTypes.array.isRequired,
  deletingShareIds: PropTypes.array.isRequired,
  awaringShareIds: PropTypes.array.isRequired,
  returningShareIds: PropTypes.array.isRequired,
  onCreateShare: PropTypes.func.isRequired,
  onDeleteShare: PropTypes.func.isRequired,
  onAwareShare: PropTypes.func.isRequired,
  onReturnShare: PropTypes.func.isRequired,
  loginUser: PropTypes.object.isRequired
};

export { InPostList };
