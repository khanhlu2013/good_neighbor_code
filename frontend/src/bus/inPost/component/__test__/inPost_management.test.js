import React from "react";
import { shallow } from "enzyme";
import InPostManagementComponent from "../inPost_management";
import LoadingIcon from "../../../../share/loadingIcon";
import { rawsToPosts } from "../../../../api/_private_api_helper";
import InPostUserHistoryList from "../inPost_userHistoryList";
import InPostList from "../inPostList";
import TabPanel from "../../../../share/style/tabPanel_style";
import InPostTabEnum from "../inPost_tabEnum";

describe("inPost_management component", () => {
  it("on did mount -> can fetch data", () => {
    const fetchInPosts = jest.fn();
    shallow(
      <InPostManagementComponent
        fetchInPosts={fetchInPosts}
        posts={[]}
        isFetchingPosts={false}
        isInitPosts={true}
        //calculated
        requestPosts={[]}
        borrowPosts={[]}
        approveAlertPosts={[]}
        returnShares={[]}
      />
    );
    expect(fetchInPosts).toHaveBeenCalledTimes(1);
  });

  it("on render -> can show loading icon during fetching", () => {
    const wrap = shallow(
      <InPostManagementComponent
        fetchInPosts={jest.fn()}
        posts={[]}
        isFetchingPosts={true}
        isInitPosts={false}
        //calculated
        requestPosts={[]}
        borrowPosts={[]}
        approveAlertPosts={[]}
        returnShares={[]}
      />
    );
    expect(wrap.find(LoadingIcon)).toHaveLength(1);
  });

  it("on render -> can show emptiness when not loading and not init posts", () => {
    const wrap = shallow(
      <InPostManagementComponent
        fetchInPosts={jest.fn()}
        posts={[]}
        isFetchingPosts={false}
        isInitPosts={false}
        //calculated
        requestPosts={[]}
        borrowPosts={[]}
        approveAlertPosts={[]}
        returnShares={[]}
      />
    );
    expect(wrap.find(LoadingIcon)).toHaveLength(0);
    expect(wrap).toMatchSnapshot();
  });

  it("on render -> can show all posts", () => {
    const fixtures = [
      {
        shares: [{}],
        user: {
          _id: "5bf40631236352000018b0eb",
          email: "my@friend.com",
          name: "My Friend"
        },
        post: {
          _id: "5bf406545407a8bb75578493",
          isActive: true,
          user: "5bf40631236352000018b0eb",
          title: "p title",
          description: "p description",
          dateCreate: "2018-11-20T13:04:20.390Z",
          __v: 0
        }
      }
    ];
    const [post] = rawsToPosts(fixtures);

    const wrap = shallow(
      <InPostManagementComponent
        fetchInPosts={jest.fn()}
        posts={[post]}
        isFetchingPosts={false}
        isInitPosts={true}
        //calculated
        requestPosts={[]}
        borrowPosts={[]}
        approveAlertPosts={[]}
        returnShares={[]}
      />
    );
    const subjectWrap = wrap.find({ listId: "inPostList-all-react" });
    expect(subjectWrap).toHaveLength(1);
    expect(subjectWrap.get(0)).toMatchSnapshot();
  });

  it("on render -> can show request posts", () => {
    const fixtures = [
      {
        shares: [{}],
        user: {
          _id: "5bf40631236352000018b0eb",
          email: "my@friend.com",
          name: "My Friend"
        },
        post: {
          _id: "5bf406545407a8bb75578493",
          isActive: true,
          user: "5bf40631236352000018b0eb",
          title: "p title",
          description: "p description",
          dateCreate: "2018-11-20T13:04:20.390Z",
          __v: 0
        }
      }
    ];
    const [post] = rawsToPosts(fixtures);

    const wrap = shallow(
      <InPostManagementComponent
        fetchInPosts={jest.fn()}
        posts={[]}
        isFetchingPosts={false}
        isInitPosts={true}
        //calculated
        requestPosts={[post]}
        borrowPosts={[]}
        approveAlertPosts={[]}
        returnShares={[]}
      />
    );
    const subjectWrap = wrap.find({ listId: "inPostList-request-react" });
    expect(subjectWrap).toHaveLength(1);
    expect(subjectWrap.get(0)).toMatchSnapshot();
  });

  it("on render -> can show approve posts", () => {
    const fixtures = [
      {
        shares: [{}],
        user: {
          _id: "5bf40631236352000018b0eb",
          email: "my@friend.com",
          name: "My Friend"
        },
        post: {
          _id: "5bf406545407a8bb75578493",
          isActive: true,
          user: "5bf40631236352000018b0eb",
          title: "p title",
          description: "p description",
          dateCreate: "2018-11-20T13:04:20.390Z",
          __v: 0
        }
      }
    ];
    const [post] = rawsToPosts(fixtures);

    const wrap = shallow(
      <InPostManagementComponent
        fetchInPosts={jest.fn()}
        posts={[]}
        isFetchingPosts={false}
        isInitPosts={true}
        //calculated
        requestPosts={[]}
        borrowPosts={[]}
        approveAlertPosts={[post]}
        returnShares={[]}
      />
    );
    const subjectWrap = wrap.find({ listId: "inPostList-approveNote-react" });
    expect(subjectWrap).toHaveLength(1);
    expect(subjectWrap.get(0)).toMatchSnapshot();
  });

  it("on render -> can show borrow posts", () => {
    const fixtures = [
      {
        shares: [{}],
        user: {
          _id: "5bf40631236352000018b0eb",
          email: "my@friend.com",
          name: "My Friend"
        },
        post: {
          _id: "5bf406545407a8bb75578493",
          isActive: true,
          user: "5bf40631236352000018b0eb",
          title: "p title",
          description: "p description",
          dateCreate: "2018-11-20T13:04:20.390Z",
          __v: 0
        }
      }
    ];
    const [post] = rawsToPosts(fixtures);

    const wrap = shallow(
      <InPostManagementComponent
        fetchInPosts={jest.fn()}
        posts={[]}
        isFetchingPosts={false}
        isInitPosts={true}
        //calculated
        requestPosts={[]}
        borrowPosts={[post]}
        approveAlertPosts={[]}
        returnShares={[]}
      />
    );
    const subjectWrap = wrap.find({ listId: "inPostList-borrow-react" });
    expect(subjectWrap).toHaveLength(1);
    expect(subjectWrap.get(0)).toMatchSnapshot();
  });

  it("on render -> can show user history list", () => {
    const fixtures = [
      {
        shares: [
          {
            _id: "5bf40b045407a8bb75578494",
            isAwareApprove: true,
            isReturn: true,
            isAwareReturn: false,
            post: "5bf406545407a8bb75578493",
            borrower: {
              _id: "5bf40631236352000018b0ed",
              email: "me@me.com",
              name: "I Myself Me"
            },
            dateCreate: "2018-11-20T13:24:20.555Z",
            __v: 0,
            isApprove: true,
            dateReturn: "2018-11-20T13:24:35.347Z"
          }
        ],
        user: {
          _id: "5bf40631236352000018b0eb",
          email: "my@friend.com",
          name: "My Friend"
        },
        post: {
          _id: "5bf406545407a8bb75578493",
          isActive: true,
          user: "5bf40631236352000018b0eb",
          title: "p title",
          description: "p description",
          dateCreate: "2018-11-20T13:04:20.390Z",
          __v: 0
        }
      }
    ];

    const [post] = rawsToPosts(fixtures);
    const [share] = post.shares;

    const wrap = shallow(
      <InPostManagementComponent
        fetchInPosts={jest.fn()}
        posts={[]}
        isFetchingPosts={false}
        isInitPosts={true}
        //calculated
        requestPosts={[]}
        borrowPosts={[]}
        approveAlertPosts={[]}
        returnShares={[share]}
      />
    );
    const subjectWrap = wrap.find(InPostUserHistoryList);
    expect(subjectWrap).toHaveLength(1);
    expect(subjectWrap.get(0)).toMatchSnapshot();
  });

  it("when render -> can hide/show TabPanel correctly", () => {
    const wrap = shallow(
      <InPostManagementComponent
        fetchInPosts={jest.fn()}
        posts={[]}
        isFetchingPosts={false}
        isInitPosts={true}
        //calculated
        requestPosts={[]}
        borrowPosts={[]}
        approveAlertPosts={[]}
        returnShares={[]}
      />
    );
    const getListIdOfShownTab = wrap => {
      const result = wrap
        .find(TabPanel)
        .find({ show: true })
        .find(InPostList);
      expect(result).toHaveLength(1);

      return result.at(0).prop("listId");
    };

    expect(getListIdOfShownTab(wrap)).toEqual("inPostList-all-react");

    wrap.setState({ selectTab: InPostTabEnum.REQUEST });
    expect(getListIdOfShownTab(wrap)).toEqual("inPostList-request-react");

    wrap.setState({ selectTab: InPostTabEnum.APPROVE });
    expect(getListIdOfShownTab(wrap)).toEqual("inPostList-approveNote-react");

    wrap.setState({ selectTab: InPostTabEnum.BORROW });
    expect(getListIdOfShownTab(wrap)).toEqual("inPostList-borrow-react");

    wrap.setState({ selectTab: InPostTabEnum.HISTORY });
    expect(
      wrap
        .find(TabPanel)
        .find({ show: true })
        .find(InPostUserHistoryList)
    ).toHaveLength(1);
  });
});
