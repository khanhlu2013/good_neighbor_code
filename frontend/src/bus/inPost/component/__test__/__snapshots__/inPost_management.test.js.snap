// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inPost_management component on render -> can show all posts 1`] = `
<InPostList
  listId="inPostList-all-react"
  posts={
    Array [
      Post {
        "dateCreate": 2018-11-20T13:04:20.390Z,
        "description": "p description",
        "id": "5bf406545407a8bb75578493",
        "isActive": true,
        "shares": Array [],
        "title": "p title",
        "user": User {
          "email": "my@friend.com",
          "id": "5bf40631236352000018b0eb",
          "name": "My Friend",
          "profileImageUrl": undefined,
        },
      },
    ]
  }
/>
`;

exports[`inPost_management component on render -> can show approve posts 1`] = `
<InPostList
  listId="inPostList-approveNote-react"
  posts={
    Array [
      Post {
        "dateCreate": 2018-11-20T13:04:20.390Z,
        "description": "p description",
        "id": "5bf406545407a8bb75578493",
        "isActive": true,
        "shares": Array [],
        "title": "p title",
        "user": User {
          "email": "my@friend.com",
          "id": "5bf40631236352000018b0eb",
          "name": "My Friend",
          "profileImageUrl": undefined,
        },
      },
    ]
  }
/>
`;

exports[`inPost_management component on render -> can show borrow posts 1`] = `
<InPostList
  listId="inPostList-borrow-react"
  posts={
    Array [
      Post {
        "dateCreate": 2018-11-20T13:04:20.390Z,
        "description": "p description",
        "id": "5bf406545407a8bb75578493",
        "isActive": true,
        "shares": Array [],
        "title": "p title",
        "user": User {
          "email": "my@friend.com",
          "id": "5bf40631236352000018b0eb",
          "name": "My Friend",
          "profileImageUrl": undefined,
        },
      },
    ]
  }
/>
`;

exports[`inPost_management component on render -> can show emptiness when not loading and not init posts 1`] = `
<div
  id="inPostManagement-react"
/>
`;

exports[`inPost_management component on render -> can show request posts 1`] = `
<InPostList
  listId="inPostList-request-react"
  posts={
    Array [
      Post {
        "dateCreate": 2018-11-20T13:04:20.390Z,
        "description": "p description",
        "id": "5bf406545407a8bb75578493",
        "isActive": true,
        "shares": Array [],
        "title": "p title",
        "user": User {
          "email": "my@friend.com",
          "id": "5bf40631236352000018b0eb",
          "name": "My Friend",
          "profileImageUrl": undefined,
        },
      },
    ]
  }
/>
`;

exports[`inPost_management component on render -> can show user history list 1`] = `
<InPostUserHistoryList
  shares={
    Array [
      Share {
        "borrower": User {
          "email": "me@me.com",
          "id": "5bf40631236352000018b0ed",
          "name": "I Myself Me",
          "profileImageUrl": undefined,
        },
        "dateCreate": 2018-11-20T13:24:20.555Z,
        "dateReturn": 2018-11-20T13:24:35.347Z,
        "id": "5bf40b045407a8bb75578494",
        "isApprove": true,
        "isAwareApprove": true,
        "isAwareReturn": false,
        "isReturn": true,
        "post": Post {
          "dateCreate": 2018-11-20T13:04:20.390Z,
          "description": "p description",
          "id": "5bf406545407a8bb75578493",
          "isActive": true,
          "shares": Array [
            [Circular],
          ],
          "title": "p title",
          "user": User {
            "email": "my@friend.com",
            "id": "5bf40631236352000018b0eb",
            "name": "My Friend",
            "profileImageUrl": undefined,
          },
        },
      },
    ]
  }
/>
`;
